src/dodal/devices/undulator_dcm.py

```diff
@@ -0,0 +1,72 @@
+import numpy as np
+from numpy import argmin, loadtxt, ndarray
+from ophyd import Component, Device, Signal
+from ophyd.status import Status
+
+from dodal.devices.DCM import DCM
+from dodal.devices.undulator import Undulator, UndulatorGapAccess
+from dodal.log import LOGGER
+
+
+class AccessError(Exception):
+    pass
+
+
+def _get_energy_distance_table(lookup_table_path: str) -> ndarray:
+    return loadtxt(lookup_table_path, comments=["#", "Units"])
+
+
+def _get_closest_gap_to_dcm_energy(
+    dcm_energy: float, energy_to_distance_table: ndarray
+) -> float:
+    table = energy_to_distance_table.transpose()
+    idx = argmin(np.abs(table[0] - dcm_energy))
+    return table[1][idx]
+
+
+# Composite device to handle changing beamline energies
+class UndulatorDCM(Device):
+    class EnergySignal(Signal):
+        def set(self, value, *, timeout=None, settle_time=None, **kwargs) -> Status:
+            access_level = self.parent.undulator.gap_access.get()
+            if access_level == UndulatorGapAccess.DISABLED.value:
+                LOGGER.error("Undulator gap access is disabled. Contact Control Room")
+                raise AccessError(
+                    "Undulator gap access is disabled. Contact Control Room"
+                )
+
+            # Get 2d np.array converting energies to undulator gap distance, from lookup table
+            energy_to_distance_table = _get_energy_distance_table(
+                self.parent.undulator.lookup_table_path
+            )
+            LOGGER.info(f"Setting DCM energy to {value:.2f} kev")
+            self.parent.dcm.energy_in_kev.move(value)
```

Comment:
Should: `move` returns a status, we should `and` all the statuses and return that on this `set`